### PR TITLE
Retire KingDanger array

### PR DESCRIPTION
**Pull request by snicolet**
Rescale the king danger variables in evaluate_king() to
suppress the KingDanger[] array. This avoids the cost of
the memory accesses to the array and simplifies the non-linear
transformation used.

Full credits to "hxim" for the seminal idea and implementation,
see pull request #786.

Passed STC:
LLR: 2.95 (-2.94,2.94) [-3.00,1.00]
Total: 9649 W: 1829 L: 1689 D: 6131

Passed LTC:
LLR: 2.95 (-2.94,2.94) [-3.00,1.00]
Total: 53494 W: 7254 L: 7178 D: 39062

Bench: 7789537

### CODE REVIEW

**Comment by DamasClasicas** [comment]
Very nice! Congratulations!!

**Comment by AlexandreMasta** [comment]
Scalability is not affected? The down curve seems too much inclined.

**Comment by mcostalba** [comment]
@snicolet Just to have a better understanding of this patch, could you please resubmit with SPRT[0, 4] at LTC? Thanks.

This is more acceptable as a simplification than hxim's original version, but the concept behind the 2 patches is very similar, so I guess it is correct to pass through similar treatment, although in this case SPRT[0, 4] result is not decisive for merging this one but can be an added element for final decision.

**Comment by snicolet** [comment]
Just for reference, here is a comparison between this patch and the 
current for the operations done at the end of evaluation_king for 
scoring the king danger.

• this patch:

a) ensure that danger is positive
b) multiply by itself
c) shift 12 bits to the right
d) ensure that result is less than two bishops
e) shift 16 bits to the left
f) add to score

• current master:

a) ensure that attack is positive
b) ensure that result is less than 399
c) shift 2 bits to the left
d) add the address of KingDanger[0]
e) read memory
f) add to score

So basically hxim's idea and this patch replaces a memory access
by a multiplication, debunks the idea that the pseudo-sigmoid
is necessary for king danger evaluation (a capped quadratic is
sufficient), avoids a memory access, suppress an array of 400
values and gives a semantics to the order of magnitude of the
king danger score, all this with 24 less lines of code: yes, I think
this qualifies as a simplification per the guidelines https://github.com/glinscott/fishtest/wiki/Creating-my-first-test

**Comment by Rocky640** [comment]
In this patch, the new kingDanger units are more or less equal to the old attackUnits * 11 - 5

a) The following graph uses the attackUnits on the x-axis, and shows the mg_score on the y axis.
We can see the plateau at 2 * BishopValueMg = 1652

2016-09-08_20h46_28

Comparing the 2 curves, near x=190, the y-delta is maximal and is -165.
This might seems a bit large if we consider that PawnValueMg is 198...
More exactly, at x=190, y-old = 1226  and y-new = 1061

The truth is that even with this lower 1061, 
the endangered king camp has already a penalty of 1061/198 = 5 pawns
so everything is probably losing for that camp at that point.

The most important part of the graph are, I think, between y = 0 and y= 2 * PawnValueMg=400
and the curves are close enough there.

b) If there is any concern, one can tune the -5 at line 426

For example, one can obtain a better match between the curves with 50 instead of -5 at line 426
With this, at x=190, y-old=1226 and y-new' = -108

2016-09-08_20h57_42

c) Finally, I'm still puzzled by this conversion by a factor of 11. 
OK, at the end, we divide by a nice 4096, and these are just numbers after all.

But we could keep our good old attackUnits values (which are not that sensitive, by the way), as in current master, and only change the final formula as follow

`int A = std::max(attackUnits, 0); 

score -= make_score( std::min( A * A /  32 , twoBishops) , 0); 
`
2016-09-08_21h00_49

**Comment by mcostalba** [comment]
If it is simpler is better, you are free to try on fishtest.

**Comment by snicolet** [comment]
@Rocky640 : thanks for the graphics and the inspiring discussion, as usual!

Concerning (c) and the multiplication of attack units by factor 11, @hxim explained that he wanted the interesting property that the slope of the curves would be close to 1 on the right part of the curve (say when y >= 600), so that in that regime where we start accumulating a winning attack on the king, one kingDanger unit would now be approximatively equivalent to one midgame Value unit in the rest of the evaluation.

This is not a bad property to have indeed, as it might make for easier reasoning about some possible compensations with the other parts of the eval, or *might* lead to further simplifications later: for instance the two variables are in the same scale in the lines like

````
kingDanger += RookCheck, score -= SafeCheck;
````

**Comment by snicolet** [comment]
Meanwhile, the SPRT[0..4] of the patch in the PR is doing badly, LLR -1.3 after 7700 games. I expected it to fail (the previous 53494 LTC games hinted at a Elo gain of 0.23 Elo, and http://chess-sprt-calc.azurewebsites.net gives a pass probability of 5% for an Elo neutral patch in SPRT[0..4]), but still the new result of -2 +-5.5 (95%) is showering a little bit my enthusiasm for that version :(

I have submitted two tests with new versions: they simplify away the difficult line 486, and try compensate this by using parabolas which are a little bit compressed on the y axis (with reasoning that we will be using this compressed parabolas using a x variable a little bit shifted to the right), and I plan to try these two versions when the framework is under a little bit less of pressure than now.

Please feel free to test other ideas too, plus on est de fous plus on rit !

**Comment by mcostalba** [comment]
Truncated test ELO is not reliable, you should not deduce ELO form an
unfinished test.

**Comment by Kingdefender** [comment]
Yes but Stephane does not truncate the test, so he does nothing wrong in my opinion. As long as his sampling points are really random points it is fine to calculate a running Elo estimate from it. SPRT itself is a truncation based on a test of beforehand indeterminate length so just your truncation rules have to be statistically sound and repeatable, not possibly biased by the tester. You could make a graph of the running Elo and the error margins as they are considered by SPRT and once the error margins cross a certain line, the test terminates.   

The Elo calculated at SPRT termination point only is not random anymore, so you get better information from a sample done at a random point (and smaller error margins when the number of games is higher, that is still true of course).

**Comment by mcostalba** [comment]
Merged with 01f2466f6eba4d4b

Slightly edited the final kingDanger score computation.
